### PR TITLE
[INFRA] introduce SEQAN3_DEPRECATED_HEADER

### PR DESCRIPTION
**Pull request by marehr**
This PR introduces SEQAN3_DEPRECATED_HEADER.

Part of https://github.com/seqan/product_backlog/issues/160 and https://github.com/seqan/product_backlog/issues/244.

---

### move header and test

```bash
mkdir -p include/seqan3/utility # maybe create target folder

cp include/seqan3/{core,utility}/math.hpp
git mv test/unit/{core,utility}/math_test.cpp # if it exists
```

* [x] move header and tests

### check new header

`include/seqan3/utility/math.hpp` has
* [x] no unused headers
* [x] include order is correct (see 4. in https://github.com/seqan/seqan3/wiki/File-structure-and-naming#contents)

```bash
edit include/seqan3/utility/math.hpp
git add include/seqan3/utility/math.hpp
```

### deprecate header

```bash
edit include/seqan3/core/math.hpp
git add include/seqan3/core/math.hpp
```

* [x] in `\file` append `\brief` with [DEPRECATED]
* [x] in `\file` add `\deprecated This header will be removed in 3.1. Please use seqan3/utility/math.hpp instead.`
* [x] delete everything after `#pragma once`
* [x] add include after `#pragma once` to new header `#include <include/seqan3/utility/math.hpp>`
* [x] add deprecation notice after `#include`

  ```c++
  SEQAN3_DEPRECATED_HEADER(
     "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/alphabet/aminoacid/translation_new.hpp> instead.")
  ```

File looks like this:

```c++
// -----------------------------------------------------------------------------------------------------
// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
// -----------------------------------------------------------------------------------------------------

/*!\file
 * \brief [DEPRECATED] Provides math related functionality.
 * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
 * \deprecated This header will be removed in 3.1. Please use seqan3/utility/math.hpp instead.
 */

#pragma once

#include <seqan3/utility/math.hpp>

SEQAN3_DEPRECATED_HEADER(
    "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/utility/math.hpp> instead.")
```

### search and replace includes

* [x] #include <seqan3/core/math.hpp> by #include <seqan3/utility/math.hpp>

```bash
git add -p # double check if change was correct
```

### edit CMakeLists.txt

```bash
edit test/unit/core/CMakeLists.txt # remove seqan3_test entry
edit test/unit/utility/CMakeLists.txt # add seqan3_test entry
git add test/unit/core/CMakeLists.txt
git add test/unit/utility/CMakeLists.txt
```

### update snippets

* [x] go through include/seqan3/utility/math.hpp (search for `\include` or `\snippet`)

In this case I got the following snippets
* `test/snippet/core/detail/floor_log2.cpp`
* `test/snippet/core/detail/ceil_log2.cpp`
* `test/snippet/core/pow.cpp`

```bash
mkdir -p test/snippet/utility
mkdir -p test/snippet/utility/detail
git mv test/snippet/{core,utility}/detail/floor_log2.cpp
git mv test/snippet/{core,utility}/detail/ceil_log2.cpp
git mv test/snippet/{core,utility}/pow.cpp

edit include/seqan3/utility/math.hpp # change snippets paths
git add -p
```

### update performance tests

* [x] update `test/unit/performance/`
  * (we only have performance tests for char_operations, simd and bit_manipulation_benchmark.cpp in core/)

### CODE REVIEW

**Review by eseiler — APPROVED** [review]

File: include/seqan3/core/platform.hpp [hunk 1]
```diff
@@ -189,9 +189,27 @@
...
 
+//!\brief _Pragma requires a string-literal and # makes it a string
+#ifndef SEQAN3_PRAGMA
+#   define SEQAN3_PRAGMA(non_string_literal) _Pragma(#non_string_literal)
+#endif
+
 //!\brief Deprecation message for SeqAn 3.1.0 release.
-#if !defined(SEQAN3_DEPRECATED_310)
-#   define SEQAN3_DEPRECATED_310 [[deprecated("This will be removed in SeqAn-3.1.0; please see the documentation.")]]
+#ifndef SEQAN3_DEPRECATED_310
```
Comment (rrahn) [inline include/seqan3/core/platform.hpp:198]: why here `ifndef` but in all other places `!defined(...)`?

File: include/seqan3/core/platform.hpp [hunk 2]
```diff
@@ -189,9 +189,27 @@
 //  Deprecation Messages
 // ============================================================================
 
+//!\brief _Pragma requires a string-literal and # makes it a string
+#ifndef SEQAN3_PRAGMA
+#   define SEQAN3_PRAGMA(non_string_literal) _Pragma(#non_string_literal)
```
Comment (rrahn) [inline include/seqan3/core/platform.hpp:194]: _Pragma is portable between at least gcc and clang?

**Review by rrahn — COMMENTED** [review]
Looks very good. Thank you. Two open questions though.

File: include/seqan3/core/platform.hpp (hunk 2 above)
Comment (marehr) [reply include/seqan3/core/platform.hpp:194]: yes

File: include/seqan3/core/platform.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/core/platform.hpp:198]: most places are `#ifndef` or `#ifdef`.

Comment (rrahn) [reply include/seqan3/core/platform.hpp:198]: but not the line below.

File: include/seqan3/core/platform.hpp [hunk 3]
```diff
@@ -189,9 +189,27 @@
...
+#endif
+
 //!\brief Deprecation message for SeqAn 3.1.0 release.
-#if !defined(SEQAN3_DEPRECATED_310)
-#   define SEQAN3_DEPRECATED_310 [[deprecated("This will be removed in SeqAn-3.1.0; please see the documentation.")]]
+#ifndef SEQAN3_DEPRECATED_310
+#   if !defined(SEQAN3_DISABLE_DEPRECATED_WARNINGS)
```
Comment (rrahn) [inline include/seqan3/core/platform.hpp:199]:
```suggestion
#   ifndef SEQAN3_DISABLE_DEPRECATED_WARNINGS
```

File: include/seqan3/core/platform.hpp [hunk 4]
```diff
@@ -189,9 +189,27 @@
...
+#       define SEQAN3_DEPRECATED_310 [[deprecated("This will be removed in SeqAn-3.1.0; please see the documentation.")]]
+#   else
+#       define SEQAN3_DEPRECATED_310 /**/
+#   endif
+#endif
+
+//!\brief Deprecation message for deprecated header.
+#ifndef SEQAN3_DEPRECATED_HEADER
+#   if !defined(SEQAN3_DISABLE_DEPRECATED_WARNINGS)
```
Comment (rrahn) [inline include/seqan3/core/platform.hpp:208]:
```suggestion
#   ifndef SEQAN3_DISABLE_DEPRECATED_WARNINGS
```

File: include/seqan3/core/platform.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/core/platform.hpp:198]: ah sorry I misunderstood you, this is a leftover from a rewrite, before I had both conditions on one line

**Review by rrahn — APPROVED** [review]

**Pull request merged**